core/src/machine/step.rs

```diff
@@ -80,6 +79,79 @@ impl Machine {
             }}
         }
 
+        /// This is a template macro for all SBC instructions. Input should be a [`Byte`].
+        macro_rules! sbc {
+            ($x:expr) => {{
+                let val = $x - (self.cpu.carry() as u8);
+                sub!(val)
+            }}
+        }
+
+        /// This is a template macro for all ADD A, b instructions (where `b` should be a [`Byte`]).
+        macro_rules! add {
+            ($x:expr) => {{
+                let (carry, half_carry) = self.cpu.a.add_with_carries($x);
+                let zero = self.cpu.a == Byte::zero();
+                set_flags!(self.cpu.f => zero 0 half_carry carry);
+
+                false
+            }}
+        }
+
+        /// This is a template macro for all ADC A, b instructions (where `b` should be a [`Byte`]).
+        macro_rules! adc {
+            ($x:expr) => {{
+                let val = $x + (self.cpu.carry() as u8);
+                add!(val)
+            }}
+        }
+
+        /// This is a template macro for all AND b instructions (where `b` should be a [`Byte`]).
+        macro_rules! and {
+            ($x:expr) => {{
+                self.cpu.a &= $x;
+                let zero = self.cpu.a == Byte::zero();
+                set_flags!(self.cpu.f => zero 0 1 0);
+
+                false
+            }}
+        }
+
+        /// This is a template macro for all XOR b instructions (where `b` should be a [`Byte`]).
+        macro_rules! xor {
+            ($x:expr) => {{
+                self.cpu.a ^= $x;
+                set_flags!(self.cpu.f => 1 0 0 0);
```

Comment:
Zero flag is wrong here: it is set like in `and`. XOR with another register doesn't need to lead to a=0.